src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs

```diff
@@ -118,237 +134,156 @@ public string GetNode()
             return node?.NodeTag;
         }
 
-        private void ReplicateToDestination()
+        private void PullReplication()
         {
-            try
+            NativeMemory.EnsureRegistered();
+
+            AddReplicationPulse(ReplicationPulseDirection.OutgoingInitiate);
+            if (_log.IsInfoEnabled)
+                _log.Info($"Pull replicate to {Destination.FromString()}");
+
+            using (_stream)
+            using (_interruptibleRead = new InterruptibleRead(_database.DocumentsStorage.ContextPool, _stream))
+            using (_buffer = JsonOperationContext.ManagedPinnedBuffer.LongLivedInstance())
             {
-                AddReplicationPulse(ReplicationPulseDirection.OutgoingInitiate);
-                NativeMemory.EnsureRegistered();
-                if (_log.IsInfoEnabled)
-                    _log.Info($"Will replicate to {Destination.FromString()} via {_connectionInfo.Url}");
+                InitialHandshake();
+                Replicate();
+            }
+        }
 
-                using (_parent._server.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
-                using (context.OpenReadTransaction())
-                {
-                    var record = _parent.LoadDatabaseRecord();
-                    if (record == null)
-                        throw new InvalidOperationException($"The database record for {_parent.Database.Name} does not exist?!");
+        private void Replication()
+        {
+            NativeMemory.EnsureRegistered();
 
-                    if (record.Encrypted && Destination.Url.StartsWith("https:", StringComparison.OrdinalIgnoreCase) == false)
-                        throw new InvalidOperationException(
-                            $"{record.DatabaseName} is encrypted, and require HTTPS for replication, but had endpoint with url {Destination.Url} to database {Destination.Database}");
-                }
+            var certificate = _parent.GetCertificateForReplication(Destination, out var authorizationInfo);
 
-                var task = TcpUtils.ConnectSocketAsync(_connectionInfo, _parent._server.Engine.TcpConnectionTimeout, _log);
-                task.Wait(CancellationToken);
-                using (Interlocked.Exchange(ref _tcpClient, task.Result))
-                {
-                    var wrapSsl = TcpUtils.WrapStreamWithSslAsync(_tcpClient, _connectionInfo, _parent._server.Server.Certificate.Certificate,
-                        _parent._server.Engine.TcpConnectionTimeout);
-                    wrapSsl.Wait(CancellationToken);
+            using (_parent._server.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
+            {
+                var record = _parent.LoadDatabaseRecord();
```

Comment:
If this happens often, then we can use ReadRawRecord?

Reply:
Probably a good idea